File: banchus-app/src/main/java/pe/nanamochi/banchus/protocol/handlers/MatchScoreUpdateHandler.java
package pe.nanamochi.banchus.protocol.handlers;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import pe.nanamochi.banchus.core.Packets;
import pe.nanamochi.banchus.database.entity.Session;
import pe.nanamochi.banchus.domain.enums.SlotStatus;
import pe.nanamochi.banchus.packets.client.MatchScoreUpdatePacket;
import pe.nanamochi.banchus.protocol.AbstractPacketHandler;
import pe.nanamochi.banchus.protocol.HandleClientPacket;
import pe.nanamochi.banchus.protocol.PacketWriter;
import pe.nanamochi.banchus.service.MatchBroadcastService;
import pe.nanamochi.banchus.service.MultiplayerService;

@Slf4j
@Component
@RequiredArgsConstructor
@HandleClientPacket(Packets.OSU_MATCH_SCORE_UPDATE)
public class MatchScoreUpdateHandler extends AbstractPacketHandler<MatchScoreUpdatePacket> {
  private final MultiplayerService multiplayerService;
  private final MatchBroadcastService matchBroadcastService;
  private final PacketWriter packetWriter;

  @Override
  public void handle(
      MatchScoreUpdatePacket packet, Session session, ByteArrayOutputStream responseStream)
      throws IOException {
    if (session.getMultiplayerMatchId() == null) {
      log.warn(
          "User {} sent a match score frame but they are not in a match.",
          session.getUser().getUsername());
      return;
    }

    multiplayerService
        .findById(session.getMultiplayerMatchId())
        .ifPresentOrElse(
            match ->
                multiplayerService
                    .findSlotBySessionId(match.getMatchId(), session.getId())
                    .ifPresentOrElse(
                        slot -> {
                          packet.getFrame().setId(slot.getSlotId());

                          try {
                            ByteArrayOutputStream packetStream = new ByteArrayOutputStream();
                            packetWriter.writePacket(
                                packetStream,
                                new pe.nanamochi.banchus.packets.server.MatchScoreUpdatePacket(
                                    packet.getFrame()));

                            matchBroadcastService.broadcastToMatch(
                                match, packetStream.toByteArray(), SlotStatus.PLAYING.getValue());
                          } catch (IOException e) {
                            log.error("Error broadcasting score update", e);
                          }
                        },
                        () ->
                            log.warn(
                                "User {} sent a match score frame but they are not in a slot.",
                                session.getUser().getUsername())),
            () ->
                log.warn(
                    "User {} sent a match score frame but their match does not exist.",
                    session.getUser().getUsername()));
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/protocol/handlers/MatchScoreUpdateHandler.kt
package pe.nanamochi.banchus.protocol.handlers

import com.github.michaelbull.result.onFailure
import java.io.ByteArrayOutputStream
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import pe.nanamochi.banchus.core.PacketType
import pe.nanamochi.banchus.database.entity.Session
import pe.nanamochi.banchus.infrastructure.protocol.HandleClientPacket
import pe.nanamochi.banchus.packets.client.MatchScoreUpdatePacket
import pe.nanamochi.banchus.protocol.AbstractPacketHandler
import pe.nanamochi.banchus.service.MultiplayerService

@Component
@HandleClientPacket(type = PacketType.OSU_MATCH_SCORE_UPDATE)
class MatchScoreUpdateHandler(private val multiplayerService: MultiplayerService) :
    AbstractPacketHandler<MatchScoreUpdatePacket>(PacketType.OSU_MATCH_SCORE_UPDATE) {
    private val log = LoggerFactory.getLogger(javaClass)

    override fun handle(
        packet: MatchScoreUpdatePacket,
        session: Session,
        responseStream: ByteArrayOutputStream,
    ) {
        multiplayerService.handleScoreUpdated(session, packet).onFailure { error ->
            log.warn("Failed to handle score update for user {}: {}", session.user?.username, error)
        }
    }
}
